README.md
# Prove
[![Hex.pm: version](https://img.shields.io/hexpm/v/prove.svg?style=flat-square)](https://hex.pm/packages/prove)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg?style=flat-square)](https://github.com/hrzndhrn/prove/blob/main/LICENSE.md)


Prove provides the macros `prove` and `batch` to write simple tests shorter.

For usage and examples, see the [documentation](https://hexdocs.pm/prove/Prove.html).


CHANGELOG.md
# Changelog

## 0.1.7 - 2021/10/28

+ Use right stacktrace.

## 0.1.6 - 2021/10/24

+ Use right stacktrace.

## 0.1.5 - 2021/10/23

+ Use right stacktrace.

## 0.1.4 - 2023/02/14

+ Minor fixes.

## 0.1.3 - 2021/08/07

+ Minor fixes.

## 0.1.2 - 2021/07/31

+ Minor fixes.

## 0.1.1 - 2021/07/22

+ Minor fixes.

## 0.1.0 - 2021/07/04

+ The very first version.



test/num_test.exs
defmodule NumTest do
  use ExUnit.Case

  import Prove

  defmodule Num do
    def check(0), do: :zero

    def check(x) when is_integer(x) do
      case rem(x, 2) do
        0 -> :even
        1 -> :odd
      end
    end

    def check(_), do: :error
  end

  describe "check/1" do
    prove Num.check(0) == :zero

    batch "returns :odd or :even" do
      prove Num.check(1) == :odd
      prove Num.check(2) == :even
      prove "for big num", Num.check(2_000) == :even
    end

    batch "returns :error" do
      prove Num.check("1") == :error
      prove Num.check(nil) == :error
    end
  end

  # The example above with regular `test`s:
  # describe "check/1" do
  #   test "returns :zero" do
  #     assert Num.check(0) == :zero
  #   end

  #   test "returns :odd or :even" do
  #     assert Num.check(1) == :odd
  #     assert Num.check(2) == :even
  #     assert Num.check(2_000) == :even
  #   end

  #   test "returns :error" do
  #     assert Num.check("1") == :error
  #     assert Num.check(nil) == :error
  #   end
  # end
end


test/prove_test.exs
defmodule ProveTest do
  use ExUnit.Case

  import Prove

  defmodule Num do
    def check(0), do: :zero
    def check(neg) when neg < 0, do: :neg

    def check(num) when is_integer(num) do
      case rem(num, 2) do
        0 -> :even
        1 -> :odd
      end
    end

    def check(_), do: :error
  end

  defp identity(x), do: x

  describe "check/1:" do
    prove Num.check(-1) == :neg
    prove Num.check(0) == :zero
    prove Num.check(1) == :odd
    prove Num.check(2) == :even
    prove "big-num:", Num.check(999_999_999) == :odd
  end

  describe "prove with different operators" do
    prove identity(5) == 5
    prove identity(5) != 1
    prove identity(5) === 5
    prove identity(5) !== 1
    prove identity(5) > 1
    prove identity(5) >= 5
    prove identity(5) < 10
    prove identity(5) <= 5
    prove "abcd" =~ "bc"
  end

  describe "batch/2" do
    batch "all together now" do
      prove identity(5) > 1
      prove identity(5) > 2
      prove identity(5) > 3
      prove "can I have a little more?", identity(5) > 4
    end

    batch "one" do
      prove identity(5) > 1
    end
  end

  # TODO
  # test "context" do
  #   assert __MODULE__.__info__(:functions) == [
  #            __ex_unit__: 0,
  #            __ex_unit__: 2,
  #            "test batch/2 all together now can I have a little more? identity(5) > 4": 1,
  #            "test batch/2 all together now identity(5) > 1": 1,
  #            "test batch/2 all together now identity(5) > 2": 1,
  #            "test batch/2 all together now identity(5) > 3": 1,
  #            "test batch/2 one identity(5) > 1": 1,
  #            "test check/1: Num.check(-1) == :neg": 1,
  #            "test check/1: Num.check(0) == :zero": 1,
  #            "test check/1: Num.check(1) == :odd": 1,
  #            "test check/1: Num.check(2) == :even": 1,
  #            "test check/1: big-num: Num.check(999999999) == :odd": 1,
  #            "test context": 1,
  #            "test prove with different operators \"abcd\" =~ \"bc\"": 1,
  #            "test prove with different operators identity(5) != 1": 1,
  #            "test prove with different operators identity(5) !== 1": 1,
  #            "test prove with different operators identity(5) < 10": 1,
  #            "test prove with different operators identity(5) <= 5": 1,
  #            "test prove with different operators identity(5) == 5": 1,
  #            "test prove with different operators identity(5) === 5": 1,
  #            "test prove with different operators identity(5) > 1": 1,
  #            "test prove with different operators identity(5) >= 5": 1
  #          ]
  # end
end


test/option_parser_test.exs
defmodule OptionParserTest do
  use ExUnit.Case

  import Prove

  describe "next" do
    def config, do: [strict: [str: :string, int: :integer, bool: :boolean]]

    batch "with strict good options" do
      prove OptionParser.next(["--str", "hello", "..."], config()) ==
              {:ok, :str, "hello", ["..."]}

      prove OptionParser.next(["--int=13", "..."], config()) == {:ok, :int, 13, ["..."]}
      prove OptionParser.next(["--bool=false", "..."], config()) == {:ok, :bool, false, ["..."]}
      prove OptionParser.next(["--no-bool", "..."], config()) == {:ok, :bool, false, ["..."]}
      prove OptionParser.next(["--bool", "..."], config()) == {:ok, :bool, true, ["..."]}
      prove OptionParser.next(["..."], config()) == {:error, ["..."]}
    end
  end
end


mix.exs
defmodule Prove.MixProject do
  use Mix.Project

  @github "https://github.com/hrzndhrn/prove"
  @version "0.1.7"
  @description "Prove provides the macros `prove` and `batch` to write simple tests shorter."

  def project do
    [
      app: :prove,
      version: @version,
      elixir: "~> 1.11",
      name: "Prove",
      description: @description,
      source_url: @github,
      start_permanent: Mix.env() == :prod,
      docs: docs(),
      package: package(),
      deps: deps()
    ]
  end

  def description do
    "Prove provides the macros `prove` and `batch` to write simple tests shorter."
  end

  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp docs do
    [
      main: "Prove",
      source_ref: "v#{@version}",
      formatters: ["html"]
    ]
  end

  defp package do
    [
      maintainers: ["Marcus Kruse"],
      licenses: ["MIT"],
      links: %{"GitHub" => @github}
    ]
  end

  defp deps do
    [
      {:ex_doc, "~> 0.25", only: :dev, runtime: false}
    ]
  end
end


lib/prove.ex
defmodule Prove do
  @moduledoc """
  Prove provides the macros `prove` and `batch` to write simple tests in `ExUnit`
  shorter.

  A `prove` is just helpful for elementary tests. Prove generates one test with
  one assert for every `prove`.

  The disadvantage of these macros is that the tests are containing fewer
  descriptions. For this reason and also if a `prove` looks too complicated, a
  regular `test` is to prefer.

  ## Example

  ```elixir
  defmodule NumTest do
    use ExUnit.Case

    import Prove

    defmodule Num do
      def check(0), do: :zero

      def check(x) when is_integer(x) do
        case rem(x, 2) do
          0 -> :even
          1 -> :odd
        end
      end

      def check(_), do: :error
    end

    describe "check/1" do
      prove Num.check(0) == :zero

      batch "returns :odd or :even" do
        prove Num.check(1) == :odd
        prove Num.check(2) == :even
        prove "for big num", Num.check(2_000) == :even
      end

      batch "returns :error" do
        prove Num.check("1") == :error
        prove Num.check(nil) == :error
      end
    end
  end
  ```
  The example above generates the following tests:
  ```shell
  $> mix test test/num_test.exs --trace --seed 0

  NumTest [test/num_test.exs]
    * prove check/1 (1) (0.00ms) [L#20]
    * prove check/1 returns :odd or :even (1) (0.00ms) [L#23]
    * prove check/1 returns :odd or :even (2) (0.00ms) [L#24]
    * prove check/1 returns :odd or :even for big num (1) (0.00ms) [L#25]
    * prove check/1 returns :error (1) (0.00ms) [L#29]
    * prove check/1 returns :error (2) (0.00ms) [L#30]


  Finished in 0.08 seconds (0.00s async, 0.08s sync)
  6 proves, 0 failures

  Randomized with seed 0
  ```

  The benefit of `prove` is that tests with multiple asserts can be avoided.
  The example above with regular `test`s:
  ```elixir
  ...
    describe "check/1" do
      test "returns :zero" do
        assert Num.check(0) == :zero
      end

      test "returns :odd or :even" do
        assert Num.check(1) == :odd
        assert Num.check(2) == :even
        assert Num.check(2_000) == :even
      end

      test "returns :error" do
        assert Num.check("1") == :error
        assert Num.check(nil) == :error
      end
    end
  ...
  ```
  ```shell
  $> mix test test/num_test.exs --trace --seed 0

  NumTest [test/num_test.exs]
    * test check/1 returns :zero (0.00ms) [L#36]
    * test check/1 returns :odd or :even (0.00ms) [L#40]
    * test check/1 returns :error (0.00ms) [L#46]


  Finished in 0.03 seconds (0.00s async, 0.03s sync)
  3 tests, 0 failures

  Randomized with seed 0
  ```

  """

  @operator [:==, :<, :>, :<=, :>=, :===, :=~, :!==, :!=, :in]

  @doc """
  A macro to write simple a simple test shorter.

  Code like:
  ```elxir
  prove identity(5) == 5
  prove identity(6) > 5
  prove "check:", identity(7) == 7
  ```
  is equivalent to:
  ```elixir
  test "(1)" do
    assert identity(5) == 5
  end

  test "(2)" do
    assert identity(6) > 5
  end

  test "check: (1)" do
    assert identity(7) == 7
  end
  ```

  `prove` supports the operators `==`, `!=`, `===`, `!==`, `<`, `<=`, `>`, `>=`,
  and `=~`.
  """
  defmacro prove(description \\ "", expr)

  defmacro prove(description, {operator, _, [_, _]} = expr)
           when is_binary(description) and operator in @operator do
    quote_prove(
      update_description(description, __CALLER__),
      expr,
      __CALLER__
    )
  end

  defmacro prove(_description, expr) do
    raise ArgumentError, message: "Unsupported do: #{Macro.to_string(expr)}"
  end

  @doc """
  Creates a batch of proves.

  A batch adds the `description` to every `prove`. This can be used to
  group`proves`s with a context. Every prove is still an own `test`.

  Code like:
  ```
  batch "valid" do
    prove 1 == 1
    prove "really", 2 == 2
  end
  ```
  is equivalent to:
  ```
  test "valid (1)" do
    assert 1 == 1
  end

  test "valid really (1)" do
    assert 2 == 2
  end
  ```
  """
  defmacro batch(description, do: {:__block__, _meta, block}) do
    {:__block__, [], quote_block(description, block, __CALLER__)}
  end

  defmacro batch(description, do: block) when is_tuple(block) do
    {:__block__, [], quote_block(description, [block], __CALLER__)}
  end

  defp quote_block(description, block, caller) do
    Enum.map(block, fn
      {:prove, meta, [op]} ->
        quote_block_prove(description, op, meta)

      {:prove, meta, [prove_description, op]} ->
        quote_block_prove("#{description} #{prove_description}", op, meta)

      _error ->
        raise CompileError,
          file: caller.file,
          line: caller.line,
          description: "A batch can only contain prove/1/2 functions"
    end)
  end

  defp quote_block_prove(description, op, meta) do
    {marker, _meta, children} =
      quote do
        prove unquote(description), unquote(op)
      end

    {marker, meta, children}
  end

  defp quote_prove(
         description,
         {operator, _meta, [_, _]} = expr,
         %{module: mod, file: file, line: line}
       )
       when is_binary(description) and operator in @operator do
    assertion = quote_assertion(expr)

    quote generated: true,
          file: file,
          line: line,
          bind_quoted: [
            assertion: Macro.escape(assertion),
            description: description,
            file: file,
            line: line,
            mod: mod
          ] do
      name = ExUnit.Case.register_test(mod, file, line, :prove, description, [])

      def unquote(name)(_) do
        unquote(assertion)
      rescue
        error in [ExUnit.AssertionError] ->
          reraise(error, __STACKTRACE__)
      end
    end
  end

  defp quote_assertion({operator, _meta, [left, right]} = expr) do
    expr = Macro.escape(expr)

    quote do
      left = unquote(left)
      right = unquote(right)

      ExUnit.Assertions.assert(Kernel.apply(Kernel, unquote(operator), [left, right]),
        left: left,
        right: right,
        expr: unquote(expr),
        message: "Prove with #{to_string(unquote(operator))} failed"
      )
    end
  end

  defp update_description(description, caller) do
    case Module.get_attribute(caller.module, :prove_counter) do
      nil ->
        Module.register_attribute(caller.module, :count, persist: false)
        Module.put_attribute(caller.module, :prove_counter, Map.put(%{}, description, 1))
        join(description, 1)

      %{^description => value} = map ->
        inc = value + 1
        Module.put_attribute(caller.module, :prove_counter, Map.put(map, description, inc))
        join(description, inc)

      map ->
        Module.put_attribute(caller.module, :prove_counter, Map.put(map, description, 1))
        join(description, 1)
    end
  end

  defp join("", b), do: "(#{b})"
  defp join(a, b), do: "#{a} (#{b})"
end
